Application/ControlMsrE2/ControlMsrE2.c

```diff
@@ -0,0 +1,210 @@
+ are licensed and made available under the terms and conditions of the BSD License
+ which accompanies this distribution.  The full text of the license may be found at
+ http://opensource.org/licenses/bsd-license.php
+ 
+ THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+ WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+ 
+ **/
+#include "ControlMsrE2.h"
+
+#define CONTEXTS_MAX 8
+
+EFI_GUID FormSetupClassGuid = EFI_HII_PLATFORM_SETUP_FORMSET_GUID;
+
+EFI_STATUS
+EFIAPI
+UefiMain (
+          IN EFI_HANDLE        ImageHandle,
+          IN EFI_SYSTEM_TABLE  *SystemTable
+          )
+{
+    EFI_HII_HANDLE * handles;
+    UINT32           OptionsCount;
+    UINT32           ContextsCount;
+    OneOfContext     Contexts[CONTEXTS_MAX];
+    
+    EFI_STATUS Status = InterpretArguments();
+    
+    if (!EFI_ERROR(Status)) {
+        Status = VerifyMSRE2 (ImageHandle, SystemTable);
+    }
+    
+    if (!EFI_ERROR(Status)) {
+        Print(L"\nBIOS Options:\n");
+
+        EFI_STRING SearchString = AllocateStrFromAscii ("cfg");
+        
+        if (IS_INTERACTIVE()) {
+            SearchString = ModifySearchString(SearchString);
+        }
+        
+        handles = HiiGetHiiHandles (NULL);
+        
+        if (handles == NULL) {
+            Print (L"Could not retrieve HiiHandles.\n");
+            Status = EFI_OUT_OF_RESOURCES;
+        }
+        else {
+            UINT32 ListHeaderCount;
+            UINT32 ListHeaderIndex;
+            
+            ListHeaderCount = 0;
+            for (EFI_HII_HANDLE* h = (EFI_HII_HANDLE*) handles; *h != NULL; h++) {
+                ListHeaderCount++;
+            }
+
+            // Keep list alive 'til program finishes.
+            // So that all lists can be searched, the results be displayed together.
+            // And from all those one Option will be selected to be changed
+            EFI_HII_PACKAGE_LIST_HEADER *ListHeaders[ListHeaderCount];
+            
+            OptionsCount = 0;
+            ContextsCount = 0;
+            ListHeaderIndex = 0;
+            
+            // For Each Handle
+            for (EFI_HII_HANDLE* h = (EFI_HII_HANDLE*) handles; (*h != NULL) && (ContextsCount < CONTEXTS_MAX); h++, ListHeaderIndex++) {
+                
+                ListHeaders[ListHeaderIndex] = HiiExportPackageLists(*h);
+                
+                if (ListHeaders[ListHeaderIndex]) {
+                    
+                    if (IS_VERBOSE()) {
+                        Print (L"Package List: ");
+                        PrintGuid (&ListHeaders[ListHeaderIndex]->PackageListGuid);
+                        Print (L"\n");
+                    }
+                    
+                    // First package in list
+                    EFI_HII_PACKAGE_HEADER* PkgHeader = PADD(ListHeaders[ListHeaderIndex], sizeof(EFI_HII_PACKAGE_LIST_HEADER));
+                    
+                    // For each package in list
+                    do {
```

Comment:
Maybe make this `while (ContextsCount < CONTEXTS_MAX || PkgHeader->Type != EFI_HII_PACKAGE_END)`?

Reply:
Could do that, but I wanted the packages to be handled like in a switch, so it is obvious that this is a  package like all the others but which marks the end of the list.